### PR TITLE
Fix unresolved arguments using typing.Annotated

### LINKED ISSUE #2855: Annotated not working properly when using `Info` and future annotations?
This is a bit of a weird one, the following code snippet

```python
from __future__ import annotations
import strawberry
import uuid
from typing_extensions import Annotated
from strawberry.types import Info


@strawberry.type
class Query:
    @strawberry.field
    def get_testing(
        self,
        info: Info[None, None],
        id_: Annotated[uuid.UUID, strawberry.argument(name="id")],
    ) -> str | None:
        return None


schema = strawberry.Schema(query=Query)

print(schema)
```

prints this:

```graphql
type Query {
  getTesting(id_: UUID!): String # ⬅️ see `id_`
}

scalar UUID
```

but if you remove info, it prints correctly:

```graphql
type Query {
  getTesting(id: UUID!): String 
}

scalar UUID
```

### PR DESCRIPTION
**Pull request by skilkis**
Reserved argument parsing, which interally also resolves annotations, exits early after detecting a reserved argument like an ``info`` argumnent. As a result, any subsequent annotations are not reserved arguments are unresolved. This results in `StrawberryArgument` not being able to determine that an argument annotation makes use of `typing.Annotated`.

This behavior is fixed by by ensuring that `StrawberryArgument` makes use of the new `StrawberryAnnotation.evaluate` method instead of consuming the raw annotation from it.

An added benefit of this fix is that by moving annotation resolving code from `StrawberryResolver` to `StrawberryAnnotation`, it makes it possible for `StrawberryArgument` and `StrawberryResolver` to share the same type evaluation cache.

Summary of Changes:
- Move type evaluation and caching logic to `StrawberryAnnotation`
- Separate type evalution from resolution with the new `StrawberryAnnotation.evaluate` method. This allows `StrawberryArgument` to access the evaluated annotation before it gets converted to `StraberryType` specializations
- Remove type evaluation cache from `StrawberryResolver` and move it to `StrawberryAnnotation`.
- Fix typing issues in `resolver.py`

Closes #2855

## Types of Changes

- [x] Core
- [x] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Documentation

## Issues Fixed or Closed by This PR

* #2855

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by botberry** [comment]
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/), so make sure to pick the appropriate type. If in doubt feel free to ask :)


Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to San Kilkis for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```

File: strawberry/types/fields/resolver.py [hunk 1]
```diff
@@ -190,6 +187,18 @@ def __call__(self, *args: str, **kwargs: Any) -> T:
     def signature(self) -> inspect.Signature:
         return Signature.from_callable(self._unbound_wrapped_func, follow_wrapped=True)
 
+    # TODO: find better name
```
Comment (skilkis) [inline strawberry/types/fields/resolver.py:190]: Not sure about this name, seems a bit silly to have `annotations` and `strawberry_annotations` even though `strawberry_annotations` includes reserved parameters as well unlike `annotations`. Could we consider removing the `annotations` property below @patrick91 ? Any ideas welcome!

File: strawberry/annotation.py [hunk 2]
```diff
@@ -249,7 +267,7 @@ def _is_optional(cls, annotation: Any) -> bool:
     def _is_list(cls, annotation: Any) -> bool:
         """Returns True if annotation is a List"""
 
-        annotation_origin = getattr(annotation, "__origin__", None)
+        annotation_origin = get_origin(annotation)
```
Comment (skilkis) [inline strawberry/annotation.py:270]: Also @patrick91 was there a reason such as performance to get these attributes from the annotations directly instead of using `get_origin`, `get_args` ? If so I can revert these changes quickly

File: strawberry/types/fields/resolver.py [hunk 3]
```diff
@@ -94,27 +100,23 @@ class ReservedType(NamedTuple):
     """
 
     name: str
-    type: Type
+    type: Type[Any]
```
Comment (skilkis) [inline strawberry/types/fields/resolver.py:103]: For consistency I'll also type these as `type` instead of `Type[Any]`

File: strawberry/annotation.py [hunk 4]
```diff
@@ -55,12 +52,19 @@
 
 
 class StrawberryAnnotation:
+    __slots__ = "annotation", "namespace", "__eval_cache__"
```
Comment (patrick91) [inline strawberry/annotation.py:55]: 🎰

File: strawberry/annotation.py [hunk 5]
```diff
@@ -81,18 +85,29 @@ def from_annotation(
...
         annotation = self.annotation
         if isinstance(annotation, str):
             annotation = ForwardRef(annotation)
 
-        args = []
-
         evaled_type = eval_type(annotation, self.namespace, None)
 
+        self.__eval_cache__ = evaled_type
+        return evaled_type
```
Comment (patrick91) [inline strawberry/annotation.py:101]: couldn't we use cached property here?

File: strawberry/annotation.py [hunk 6]
```diff
@@ -131,6 +146,8 @@ def set_namespace_from_field(self, field: StrawberryField) -> None:
         module = sys.modules[field.origin.__module__]
         self.namespace = module.__dict__
 
+        self.__eval_cache__ = None  # Invalidate cache to allow re-evaluation
```
Comment (patrick91) [inline strawberry/annotation.py:149]: why is this needed?

File: strawberry/annotation.py (hunk 2 above)
Comment (patrick91) [reply strawberry/annotation.py:270]: no, I don't think there was a particular reason 😊

File: strawberry/arguments.py [hunk 7]
```diff
@@ -94,48 +89,46 @@ def __init__(
             _deprecated_UNSET if default is inspect.Parameter.empty else default
         )
 
-        if self._annotation_is_annotated(type_annotation):
-            self._parse_annotated()
+        try:
+            evaled_type = type_annotation.evaluate()
+        except NameError:
+            pass
```
Comment (patrick91) [inline strawberry/arguments.py:95]: can we add a comment on when this might happen? (I'm assuming when using if TYPE_CHECKING)

File: strawberry/types/fields/resolver.py (hunk 3 above)
Comment (patrick91) [reply strawberry/types/fields/resolver.py:103]: I wish it was the default :D

File: strawberry/types/fields/resolver.py [hunk 8]
```diff
@@ -68,17 +68,23 @@ class ReservedName(NamedTuple):
     name: str
 
     def find(
-        self, parameters: Tuple[inspect.Parameter, ...], _: StrawberryResolver
+        self,
+        parameters: Tuple[inspect.Parameter, ...],
+        resolver: StrawberryResolver[Any],
     ) -> Optional[inspect.Parameter]:
+        del resolver
```
Comment (patrick91) [inline strawberry/types/fields/resolver.py:75]: why? :D

File: strawberry/types/fields/resolver.py (hunk 1 above)
Comment (patrick91) [reply strawberry/types/fields/resolver.py:190]: Could we just use annotations?

**Review by patrick91 — COMMENTED** [review]
left some comments!

**Review by patrick91 — COMMENTED** [review]
left some comments!

File: strawberry/annotation.py (hunk 2 above)
Comment (bellini666) [reply strawberry/annotation.py:270]: I think it is because `get_origin` and `get_args` were introduced on python 3.8, and probably weren't available on `typing_extensions` when this code was written.

File: strawberry/arguments.py (hunk 7 above)
Comment (bellini666) [reply strawberry/arguments.py:95]: Also, try/except with pass can be rewritten as:

```python
with contextlib.suppress(NameError):
    evaled_type = type_annotation.evaluate()
```

File: strawberry/types/fields/resolver.py [hunk 9]
```diff
@@ -190,6 +187,18 @@ def __call__(self, *args: str, **kwargs: Any) -> T:
...
 
+    # TODO: find better name
+    @cached_property
+    def strawberry_annotations(
+        self,
+    ) -> Dict[inspect.Parameter, Union[StrawberryAnnotation, None]]:
+        return {
+            p: StrawberryAnnotation(p.annotation, namespace=self._namespace)
+            if p.annotation is not inspect.Signature.empty
+            else None
```
Comment (bellini666) [inline strawberry/types/fields/resolver.py:198]: A small nit here, otherwise the alignment makes it harder to read

```suggestion
            p: (
                StrawberryAnnotation(p.annotation, namespace=self._namespace)
                if p.annotation is not inspect.Signature.empty
                else None
            )
```

File: tests/schema/test_arguments.py [hunk 10]
```diff
@@ -199,3 +200,40 @@ def hello(
...
+    }
+
+    scalar UUID
+    """
+
+    schema_a = strawberry.Schema(type_a.Query)
+    schema_b = strawberry.Schema(type_b.Query)
+
+    assert str(schema_a) == str(schema_b)
+    assert f"\n{schema_a}\n" == textwrap.dedent(expected)
```
Comment (bellini666) [inline tests/schema/test_arguments.py:239]: nit

```suggestion
    assert str(schema_a) == textwrap.dedent(expected).strip()
```

**Review by bellini666 — CHANGES_REQUESTED** [review]

File: strawberry/types/fields/resolver.py (hunk 1 above)
Comment (skilkis) [reply strawberry/types/fields/resolver.py:190]: Do you mean just get rid of this and use the annotations property below?

File: strawberry/annotation.py (hunk 6 above)
Comment (skilkis) [reply strawberry/annotation.py:149]: To keep it consistent, probably there isn't a use-case for it. But if the
field somehow changes and it lives in a different module, then the previously
evaluated type could be incorrect

File: strawberry/annotation.py (hunk 5 above)
Comment (skilkis) [reply strawberry/annotation.py:101]: We'd have to remove the `__slots__` declaration above which is something I've
been wanting to introduce to Strawberry classes for a while. Especially `Info`
as it would reduce memory usage

Comment (skilkis) [reply strawberry/annotation.py:101]: We can use cached property if you want to 😊

File: strawberry/types/fields/resolver.py (hunk 8 above)
Comment (skilkis) [reply strawberry/types/fields/resolver.py:75]: To satisfy Pyright's hatred of unused arguments 😂 It's the suggested way by PyLint to
fix unused argument warnings / errors. Can't use `_` here as it violates Liskov
since the signature of find would be different 😥

File: tests/schema/test_arguments.py (hunk 10 above)
Comment (skilkis) [reply tests/schema/test_arguments.py:239]: Thanks! Much better indeed 🚀

File: strawberry/arguments.py (hunk 7 above)
Comment (skilkis) [reply strawberry/arguments.py:95]: Good idea! I've added one now

File: strawberry/annotation.py (hunk 5 above)
Comment (patrick91) [reply strawberry/annotation.py:101]: oh, I didn't realise __slots__ don't make cachedproperty work! that's fine then

File: strawberry/types/fields/resolver.py (hunk 1 above)
Comment (patrick91) [reply strawberry/types/fields/resolver.py:189]: yeah, I was suggesting that 😊 do you think it is worth doing it?

**Review by bellini666 — APPROVED** [review]
No more comments, LGTM! :)

**Comment by skilkis** [comment]
> No more comments, LGTM! :)

Thanks @bellini666 !

File: strawberry/types/fields/resolver.py (hunk 1 above)
Comment (skilkis) [reply strawberry/types/fields/resolver.py:189]: @patrick91 I don't think it's worth it to go back to how it was implemented
previously with the resolved annotations being injected by `ReservedType`, it's a nicer flow to create the `StrawberryAnnotation` instances
first. Maybe we can consider replacing the current `annotation` property below?
It would be a breaking change though.

Comment (patrick91) [reply strawberry/types/fields/resolver.py:189]: let's do that in future, if we remember ;)

**Review by patrick91 — APPROVED** [review]
